File: cropaway/Models/Timeline.swift
//
//  Timeline.swift
//  cropaway
//

import Foundation
import CoreGraphics
import Combine

/// Represents a sequence of video clips with transitions between them
final class Timeline: Identifiable, ObservableObject, Codable {
    let id: UUID

    /// Name of this timeline/sequence
    @Published var name: String

    /// Tracks in the timeline (NEW: multi-track support)
    @Published var tracks: [Track]

    /// Ordered list of clips in the timeline (DEPRECATED: kept for backward compatibility)
    @Published var clips: [TimelineClip]

    /// Transitions between clips (indexed by afterClipIndex)
    @Published var transitions: [ClipTransition]

    /// In point for export range (nil = start of timeline)
    @Published var inPoint: Double?

    /// Out point for export range (nil = end of timeline)
    @Published var outPoint: Double?

    /// Timeline frame rate (default: 30fps)
    @Published var frameRate: Double

    /// Timeline resolution (default: .zero uses first clip's resolution)
    @Published var resolution: CGSize

    /// Creation date
    let dateCreated: Date

    /// Last modified date
    @Published var dateModified: Date

    private var cancellables = Set<AnyCancellable>()

    init(
        id: UUID = UUID(),
        name: String = "Untitled Sequence",
        tracks: [Track] = [],
        clips: [TimelineClip] = [],
        transitions: [ClipTransition] = [],
        inPoint: Double? = nil,
        outPoint: Double? = nil,
        frameRate: Double = 30.0,
        resolution: CGSize = .zero
    ) {
        self.id = id
        self.name = name
        self.tracks = tracks
        self.clips = clips
        self.transitions = transitions
        self.inPoint = inPoint
        self.outPoint = outPoint
        self.frameRate = frameRate
        self.resolution = resolution
        self.dateCreated = Date()
        self.dateModified = Date()

        setupObservers()
    }

    // MARK: - Codable

    enum CodingKeys: String, CodingKey {
        case id, name, tracks, clips, transitions, dateCreated, dateModified
        case inPoint, outPoint, frameRate, resolution
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(UUID.self, forKey: .id)
        name = try container.decode(String.self, forKey: .name)
        
        // Try to decode tracks first (new format)
        tracks = try container.decodeIfPresent([Track].self, forKey: .tracks) ?? []
        
        // Decode clips (kept for backward compatibility)
        clips = try container.decode([TimelineClip].self, forKey: .clips)
        
        transitions = try container.decode([ClipTransition].self, forKey: .transitions)
        dateCreated = try container.decode(Date.self, forKey: .dateCreated)
        dateModified = try container.decode(Date.self, forKey: .dateModified)
        
        // New properties with defaults for backward compatibility
        inPoint = try container.decodeIfPresent(Double.self, forKey: .inPoint)
        outPoint = try container.decodeIfPresent(Double.self, forKey: .outPoint)
        frameRate = try container.decodeIfPresent(Double.self, forKey: .frameRate) ?? 30.0
        
        // Decode resolution if present
        if let width = try? container.decodeIfPresent(Double.self, forKey: .resolution),
           let height = try? container.decodeIfPresent(Double.self, forKey: .resolution) {
            resolution = CGSize(width: width, height: height)
        } else {
            resolution = .zero
        }

        setupObservers()
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(name, forKey: .name)
        try container.encode(tracks, forKey: .tracks)
        try container.encode(clips, forKey: .clips)
        try container.encode(transitions, forKey: .transitions)
        try container.encode(dateCreated, forKey: .dateCreated)
        try container.encode(dateModified, forKey: .dateModified)
        try container.encodeIfPresent(inPoint, forKey: .inPoint)
        try container.encodeIfPresent(outPoint, forKey: .outPoint)
        try container.encode(frameRate, forKey: .frameRate)
        
        // Encode resolution as width/height
        if resolution != .zero {
            try container.encode(resolution.width, forKey: .resolution)
            try container.encode(resolution.height, forKey: .resolution)
        }
    }

    private func setupObservers() {
        // Mark as modified when tracks, clips or transitions change
        $tracks.sink { [weak self] _ in
            self?.dateModified = Date()
        }.store(in: &cancellables)
        
        $clips.sink { [weak self] _ in
            self?.dateModified = Date()
        }.store(in: &cancellables)

        $transitions.sink { [weak self] _ in
            self?.dateModified = Date()
        }.store(in: &cancellables)
        
        $inPoint.sink { [weak self] _ in
            self?.dateModified = Date()
        }.store(in: &cancellables)
        
        $outPoint.sink { [weak self] _ in
            self?.dateModified = Date()
        }.store(in: &cancellables)
    }

    // MARK: - Computed Properties

    /// Total duration of the timeline in seconds
    var totalDuration: Double {
        // For a simple concatenated timeline without transition overlap,
        // duration is simply the sum of all clip durations
        let clipsDuration = clips.reduce(0.0) { $0 + $1.trimmedDuration }

        // If we have transitions in the future that actually overlap,
        // we'll need to account for them. For now, transitions are
        // just markers and don't affect playback duration.
        // let overlapDuration = transitions.reduce(0.0) { $0 + $1.effectiveDuration }

        // Return clips duration, or minimum of 0.01 if no clips
        return max(0.01, clipsDuration)
    }

    /// Number of clips in the timeline
    var clipCount: Int {
        clips.count
    }

    /// Whether the timeline is empty
    var isEmpty: Bool {
        clips.isEmpty
    }

    /// Whether the timeline has multiple clips (can have transitions)
    var hasMultipleClips: Bool {
        clips.count > 1
    }
    
    // MARK: - Multi-Track Computed Properties
    
    /// All clips across all tracks (for unified access)
    var allClips: [TimelineClip] {
        tracks.flatMap { $0.clips }
    }
    
    /// Video tracks only, sorted by vertical order (top to bottom)
    var videoTracks: [Track] {
        tracks
            .filter { $0.mediaType == .video }
            .sorted { $0.verticalOrder > $1.verticalOrder }
    }
    
    /// Audio tracks only
    var audioTracks: [Track] {
        tracks
            .filter { $0.mediaType == .audio }
            .sorted { $0.verticalOrder > $1.verticalOrder }
    }
    
    /// Find a clip by ID across all tracks
    func findClip(withID id: UUID) -> TimelineClip? {
        for track in tracks {
            if let clip = track.clips.first(where: { $0.id == id }) {
                return clip
            }
        }
        return nil
    }
    
    /// Find track and index for a clip by ID
    func findClipLocation(withID id: UUID) -> (track: Track, clipIndex: Int)? {
        for track in tracks {
            if let index = track.clips.firstIndex(where: { $0.id == id }) {
                return (track, index)
            }
        }
        return nil
    }
    
    /// Export duration (respects in/out points)
    var exportDuration: Double {
        if let inPt = inPoint, let outPt = outPoint {
            return max(0, outPt - inPt)
        }
        return totalDuration
    }
    
    /// Whether in/out points are set
    var hasInOutPoints: Bool {
        inPoint != nil && outPoint != nil
    }
    
    // MARK: - Track Management
    
    /// Add a new track to the timeline
    func addTrack(_ track: Track) {
        tracks.append(track)
    }
    
    /// Create and add a new video track
    @discardableResult
    func createVideoTrack(name: String? = nil) -> Track {
        let trackNumber = videoTracks.count + 1
        let track = Track(
            name: name ?? "Video \(trackNumber)",
            mediaType: .video,
            verticalOrder: tracks.count
        )
        addTrack(track)
        return track
    }
    
    /// Create and add a new audio track
    @discardableResult
    func createAudioTrack(name: String? = nil) -> Track {
        let trackNumber = audioTracks.count + 1
        let track = Track(
            name: name ?? "Audio \(trackNumber)",
            mediaType: .audio,
            verticalOrder: tracks.count
        )
        addTrack(track)
        return track
    }
    
    /// Remove a track by ID
    @discardableResult
    func removeTrack(_ trackID: UUID) -> Track? {
        guard let index = tracks.firstIndex(where: { $0.id == trackID }) else { return nil }
        return tracks.remove(at: index)
    }
    
    /// Find a track by ID
    func track(withID id: UUID) -> Track? {
        tracks.first { $0.id == id }
    }
    
    /// Find the track containing a specific clip
    func track(containing clipID: UUID) -> Track? {
        tracks.first { track in
            track.clips.contains { $0.id == clipID }
        }
    }
    
    /// Move a track to a new vertical order position
    func moveTrack(from sourceIndex: Int, to destinationIndex: Int) {
        guard sourceIndex >= 0, sourceIndex < tracks.count,
              destinationIndex >= 0, destinationIndex < tracks.count else { return }
        let track = tracks.remove(at: sourceIndex)
        tracks.insert(track, at: destinationIndex)
        
        // Update vertical order for all tracks
        for (index, track) in tracks.enumerated() {
            track.verticalOrder = index
        }
    }

    // MARK: - Clip Management

    /// Add a clip to the end of the timeline
    func addClip(_ clip: TimelineClip) {
        clips.append(clip)
    }

    /// Add a clip from a video item
    func addClip(from videoItem: VideoItem) {
        let clip = TimelineClip(videoItem: videoItem)
        addClip(clip)
    }

    /// Insert a clip at a specific index
    func insertClip(_ clip: TimelineClip, at index: Int) {
        let safeIndex = max(0, min(clips.count, index))

        // Update transition indices for clips after the insertion point
        for transition in transitions where transition.afterClipIndex >= safeIndex {
            // Create new transition with updated index since afterClipIndex is let
            if let idx = transitions.firstIndex(where: { $0.id == transition.id }) {
                transitions[idx] = transition.copy(withNewIndex: transition.afterClipIndex + 1)
            }
        }

        clips.insert(clip, at: safeIndex)
    }

    /// Remove a clip at a specific index
    func removeClip(at index: Int) {
        guard index >= 0 && index < clips.count else { return }

        // Remove transitions that reference this clip
        transitions.removeAll { $0.afterClipIndex == index }

        // Update indices for transitions after this clip
        for (idx, transition) in transitions.enumerated() where transition.afterClipIndex > index {
            transitions[idx] = transition.copy(withNewIndex: transition.afterClipIndex - 1)
        }

        clips.remove(at: index)
    }

    /// Remove a specific clip
    func removeClip(_ clip: TimelineClip) {
        if let index = clips.firstIndex(where: { $0.id == clip.id }) {
            removeClip(at: index)
        }
    }

    /// Move a clip from one index to another
    func moveClip(from sourceIndex: Int, to destinationIndex: Int) {
        guard sourceIndex >= 0 && sourceIndex < clips.count,
              destinationIndex >= 0 && destinationIndex <= clips.count,
              sourceIndex != destinationIndex else { return }

        let clip = clips.remove(at: sourceIndex)

        // Adjust destination if moving forward
        let adjustedDestination = destinationIndex > sourceIndex ? destinationIndex - 1 : destinationIndex
        clips.insert(clip, at: adjustedDestination)

        // Update transition indices to match new clip positions
        // This requires careful index tracking without rebuilding all transitions
        updateTransitionIndicesAfterMove(from: sourceIndex, to: adjustedDestination)
    }

    /// Split a clip at a specific time within the clip
    func splitClip(at clipIndex: Int, atTime timeInClip: Double) -> Bool {
        guard clipIndex >= 0 && clipIndex < clips.count else { return false }

        let clip = clips[clipIndex]
        guard let newClip = clip.split(at: timeInClip / clip.trimmedDuration) else {
            return false
        }

        // Insert new clip after the current one
        insertClip(newClip, at: clipIndex + 1)
        return true
    }

    // MARK: - Transition Management

    /// Get the transition after a specific clip index
    func transition(afterClipIndex index: Int) -> ClipTransition? {
        transitions.first { $0.afterClipIndex == index }
    }

    /// Set the transition type after a specific clip
    func setTransitionType(afterClipIndex index: Int, type: TransitionType) {
        if let transition = transition(afterClipIndex: index) {
            transition.type = type
        }
    }

    /// Set the transition duration after a specific clip
    func setTransitionDuration(afterClipIndex index: Int, duration: Double) {
        if let transition = transition(afterClipIndex: index) {
            transition.duration = max(0.1, min(2.0, duration))
        }
    }

    /// Rebuild all transitions with default cut type
    private func updateTransitionIndicesAfterMove(from sourceIndex: Int, to destinationIndex: Int) {
        // When a clip moves, transitions need to update their indices
        // This is complex but avoids auto-creating transitions
        
        var updatedTransitions: [ClipTransition] = []
        
        for transition in transitions {
            var newIndex = transition.afterClipIndex
            
            // If transition is attached to the moved clip
            if transition.afterClipIndex == sourceIndex {
                newIndex = destinationIndex
            }
            // If transition is between source and destination
            else if sourceIndex < destinationIndex {
                // Moving forward: indices in between shift down
                if transition.afterClipIndex > sourceIndex && transition.afterClipIndex <= destinationIndex {
                    newIndex = transition.afterClipIndex - 1
                }
            } else {
                // Moving backward: indices in between shift up
                if transition.afterClipIndex >= destinationIndex && transition.afterClipIndex < sourceIndex {
                    newIndex = transition.afterClipIndex + 1
                }
            }
            
            updatedTransitions.append(transition.copy(withNewIndex: newIndex))
        }
        
        transitions = updatedTransitions
    }

    // MARK: - Time Calculations

    /// Get the clip at a specific timeline time
    func clip(at timelineTime: Double) -> (clip: TimelineClip, clipIndex: Int, timeInClip: Double)? {
        var currentTime: Double = 0

        for (index, clip) in clips.enumerated() {
            let clipDuration = clip.trimmedDuration

            // Account for transition overlap ONLY if transition exists
            if index > 0, let prevTransition = transition(afterClipIndex: index - 1) {
                let overlap = prevTransition.effectiveDuration / 2
                currentTime -= overlap
            }

            let clipEndTime = currentTime + clipDuration

            if timelineTime >= currentTime && timelineTime < clipEndTime {
                let timeInClip = timelineTime - currentTime
                return (clip, index, timeInClip)
            }

            currentTime = clipEndTime
            
            // Add gap if no transition to next clip
            if index < clips.count - 1, transition(afterClipIndex: index) == nil {
                currentTime += 0.02 // 2pt gap in time units
            }
        }

        // Return last clip if time is past the end
        if let lastClip = clips.last {
            return (lastClip, clips.count - 1, lastClip.trimmedDuration)
        }

        return nil
    }

    /// Get the start time of a clip in the timeline
    func startTime(forClipAt index: Int) -> Double {
        var time: Double = 0
        for i in 0..<index {
            time += clips[i].trimmedDuration
            if let transition = transition(afterClipIndex: i) {
                time -= transition.effectiveDuration
            } else {
                // Add gap if no transition exists
                time += 0.02
            }
        }
        return time
    }

    // MARK: - Resolution After Load

    /// Resolve all clip video item references after loading from persistence
    func resolveVideoItems(from videos: [VideoItem]) {
        // Resolve clips in deprecated clips array
        for clip in clips {
            clip.resolveVideoItem(from: videos)
        }
        
        // Resolve clips in all tracks
        for track in tracks {
            for clip in track.clips {
                clip.resolveVideoItem(from: videos)
            }
        }
    }
    
    // MARK: - Helper Methods

}

// MARK: - Equatable

extension Timeline: Equatable {
    static func == (lhs: Timeline, rhs: Timeline) -> Bool {
        lhs.id == rhs.id
    }
}

// MARK: - Hashable

extension Timeline: Hashable {
    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

File: CropawayWindows/CropawayWindows/Models/Timeline.cs
// Timeline.cs
// CropawayWindows

using System.Collections.ObjectModel;
using System.Text.Json.Serialization;
using CommunityToolkit.Mvvm.ComponentModel;

namespace CropawayWindows.Models;

/// <summary>
/// Represents a sequence of video clips with transitions between them.
/// Manages clip ordering, insertion, removal, movement, splitting, and
/// transition management. Maintains computed properties for total duration
/// and clip count.
/// </summary>
public partial class Timeline : ObservableObject, IEquatable<Timeline>
{
    /// <summary>
    /// Unique identifier for this timeline.
    /// </summary>
    public Guid Id { get; init; } = Guid.NewGuid();

    /// <summary>
    /// Display name for this timeline/sequence.
    /// </summary>
    [ObservableProperty]
    private string _name = "Untitled Sequence";

    /// <summary>
    /// Ordered list of clips in the timeline.
    /// </summary>
    [ObservableProperty]
    private ObservableCollection<TimelineClip> _clips = [];

    /// <summary>
    /// Transitions between clips, indexed by <see cref="ClipTransition.AfterClipIndex"/>.
    /// </summary>
    [ObservableProperty]
    private ObservableCollection<ClipTransition> _transitions = [];

    /// <summary>
    /// Date this timeline was created.
    /// </summary>
    public DateTime DateCreated { get; init; } = DateTime.Now;

    /// <summary>
    /// Date this timeline was last modified.
    /// </summary>
    [ObservableProperty]
    private DateTime _dateModified = DateTime.Now;

    public Timeline()
    {
    }

    public Timeline(string name, IEnumerable<TimelineClip>? clips = null)
    {
        Name = name;
        if (clips is not null)
        {
            _clips = new ObservableCollection<TimelineClip>(clips);
        }
    }

    // -- Computed properties --

    /// <summary>
    /// Total duration of the timeline in seconds. Sum of all trimmed clip durations.
    /// Returns a minimum of 0.01 to avoid zero-duration timelines.
    /// </summary>
    [JsonIgnore]
    public double TotalDuration
    {
        get
        {
            double clipsDuration = Clips.Sum(c => c.TrimmedDuration);
            return Math.Max(0.01, clipsDuration);
        }
    }

    /// <summary>
    /// Number of clips currently in the timeline.
    /// </summary>
    [JsonIgnore]
    public int ClipCount => Clips.Count;

    /// <summary>
    /// Whether the timeline contains no clips.
    /// </summary>
    [JsonIgnore]
    public bool IsEmpty => Clips.Count == 0;

    /// <summary>
    /// Whether the timeline has multiple clips (enabling transitions).
    /// </summary>
    [JsonIgnore]
    public bool HasMultipleClips => Clips.Count > 1;

    // -- Clip management --

    /// <summary>
    /// Adds a clip to the end of the timeline.
    /// </summary>
    public void AddClip(TimelineClip clip)
    {
        Clips.Add(clip);
        MarkModified();
    }

    /// <summary>
    /// Creates and adds a clip from a <see cref="VideoItem"/> to the end of the timeline.
    /// </summary>
    public void AddClip(VideoItem videoItem)
    {
        var clip = new TimelineClip(videoItem);
        AddClip(clip);
    }

    /// <summary>
    /// Inserts a clip at a specific index. Adjusts transition indices for clips
    /// after the insertion point.
    /// </summary>
    public void InsertClip(TimelineClip clip, int index)
    {
        int safeIndex = Math.Clamp(index, 0, Clips.Count);

        // Update transition indices for clips at or after the insertion point
        var updatedTransitions = Transitions
            .Select(t => t.AfterClipIndex >= safeIndex
                ? t.CopyWithNewIndex(t.AfterClipIndex + 1)
                : t)
            .ToList();

        Transitions = new ObservableCollection<ClipTransition>(updatedTransitions);
        Clips.Insert(safeIndex, clip);
        MarkModified();
    }

    /// <summary>
    /// Removes a clip at a specific index. Removes any transitions referencing
    /// this clip and adjusts indices for subsequent transitions.
    /// </summary>
    public void RemoveClip(int index)
    {
        if (index < 0 || index >= Clips.Count) return;

        // Remove transitions referencing this clip, adjust indices for later clips
        var updatedTransitions = Transitions
            .Where(t => t.AfterClipIndex != index)
            .Select(t => t.AfterClipIndex > index
                ? t.CopyWithNewIndex(t.AfterClipIndex - 1)
                : t)
            .ToList();

        Transitions = new ObservableCollection<ClipTransition>(updatedTransitions);
        Clips.RemoveAt(index);
        MarkModified();
    }

    /// <summary>
    /// Removes a specific clip from the timeline by finding its index.
    /// </summary>
    public void RemoveClip(TimelineClip clip)
    {
        int index = -1;
        for (int i = 0; i < Clips.Count; i++)
        {
            if (Clips[i].Id == clip.Id)
            {
                index = i;
                break;
            }
        }

        if (index >= 0)
        {
            RemoveClip(index);
        }
    }

    /// <summary>
    /// Moves a clip from one index to another, updating transition indices accordingly.
    /// </summary>
    public void MoveClip(int sourceIndex, int destinationIndex)
    {
        if (sourceIndex < 0 || sourceIndex >= Clips.Count ||
            destinationIndex < 0 || destinationIndex > Clips.Count ||
            sourceIndex == destinationIndex)
            return;

        var clip = Clips[sourceIndex];
        Clips.RemoveAt(sourceIndex);

        int adjustedDestination = destinationIndex > sourceIndex
            ? destinationIndex - 1
            : destinationIndex;

        Clips.Insert(adjustedDestination, clip);

        // Update transition indices
        UpdateTransitionIndicesAfterMove(sourceIndex, adjustedDestination);
        MarkModified();
    }

    /// <summary>
    /// Splits a clip at a specific time (in seconds) within the clip's trimmed duration.
    /// Returns true if the split was successful.
    /// </summary>
    public bool SplitClip(int clipIndex, double timeInClip)
    {
        if (clipIndex < 0 || clipIndex >= Clips.Count) return false;

        var clip = Clips[clipIndex];
        double normalizedPosition = clip.TrimmedDuration > 0
            ? timeInClip / clip.TrimmedDuration
            : 0.5;

        var newClip = clip.Split(normalizedPosition);
        if (newClip is null) return false;

        InsertClip(newClip, clipIndex + 1);
        return true;
    }

    // -- Transition management --

    /// <summary>
    /// Gets the transition after a specific clip index, if one exists.
    /// </summary>
    public ClipTransition? GetTransition(int afterClipIndex) =>
        Transitions.FirstOrDefault(t => t.AfterClipIndex == afterClipIndex);

    /// <summary>
    /// Sets the transition type after a specific clip.
    /// </summary>
    public void SetTransitionType(int afterClipIndex, TransitionType type)
    {
        var transition = GetTransition(afterClipIndex);
        if (transition is not null)
        {
            transition.Type = type;
            MarkModified();
        }
    }

    /// <summary>
    /// Sets the transition duration after a specific clip.
    /// </summary>
    public void SetTransitionDuration(int afterClipIndex, double duration)
    {
        var transition = GetTransition(afterClipIndex);
        if (transition is not null)
        {
            transition.Duration = Math.Clamp(duration, 0.1, 2.0);
            MarkModified();
        }
    }

    // -- Time calculations --

    /// <summary>
    /// Gets the clip at a specific timeline time, along with the clip index
    /// and the time offset within that clip.
    /// </summary>
    public (TimelineClip Clip, int ClipIndex, double TimeInClip)? GetClipAtTime(double timelineTime)
    {
        double currentTime = 0;

        for (int i = 0; i < Clips.Count; i++)
        {
            var clip = Clips[i];
            double clipDuration = clip.TrimmedDuration;

            // Account for transition overlap if a transition exists before this clip
            if (i > 0)
            {
                var prevTransition = GetTransition(i - 1);
                if (prevTransition is not null)
                {
                    currentTime -= prevTransition.EffectiveDuration / 2;
                }
            }

            double clipEndTime = currentTime + clipDuration;

            if (timelineTime >= currentTime && timelineTime < clipEndTime)
            {
                return (clip, i, timelineTime - currentTime);
            }

            currentTime = clipEndTime;

            // Add gap if no transition to next clip
            if (i < Clips.Count - 1 && GetTransition(i) is null)
            {
                currentTime += 0.02;
            }
        }

        // Return last clip if time is past the end
        if (Clips.Count > 0)
        {
            var lastClip = Clips[^1];
            return (lastClip, Clips.Count - 1, lastClip.TrimmedDuration);
        }

        return null;
    }

    /// <summary>
    /// Gets the start time of a clip (by index) within the overall timeline.
    /// </summary>
    public double GetClipStartTime(int clipIndex)
    {
        double time = 0;
        for (int i = 0; i < clipIndex && i < Clips.Count; i++)
        {
            time += Clips[i].TrimmedDuration;

            var transition = GetTransition(i);
            if (transition is not null)
            {
                time -= transition.EffectiveDuration;
            }
            else
            {
                time += 0.02; // Gap when no transition
            }
        }
        return time;
    }

    // -- Resolution after load --

    /// <summary>
    /// Resolves all clip video item references after loading from persistence.
    /// </summary>
    public void ResolveVideoItems(IEnumerable<VideoItem> videos)
    {
        var videoList = videos.ToList();
        foreach (var clip in Clips)
        {
            clip.ResolveVideoItem(videoList);
        }
    }

    // -- Private helpers --

    private void UpdateTransitionIndicesAfterMove(int sourceIndex, int destinationIndex)
    {
        var updatedTransitions = new List<ClipTransition>();

        foreach (var transition in Transitions)
        {
            int newIndex = transition.AfterClipIndex;

            if (transition.AfterClipIndex == sourceIndex)
            {
                newIndex = destinationIndex;
            }
            else if (sourceIndex < destinationIndex)
            {
                // Moving forward: indices between shift down
                if (transition.AfterClipIndex > sourceIndex &&
                    transition.AfterClipIndex <= destinationIndex)
                {
                    newIndex = transition.AfterClipIndex - 1;
                }
            }
            else
            {
                // Moving backward: indices between shift up
                if (transition.AfterClipIndex >= destinationIndex &&
                    transition.AfterClipIndex < sourceIndex)
                {
                    newIndex = transition.AfterClipIndex + 1;
                }
            }

            updatedTransitions.Add(transition.CopyWithNewIndex(newIndex));
        }

        Transitions = new ObservableCollection<ClipTransition>(updatedTransitions);
    }

    private void MarkModified()
    {
        DateModified = DateTime.Now;
    }

    public bool Equals(Timeline? other)
    {
        if (other is null) return false;
        if (ReferenceEquals(this, other)) return true;
        return Id == other.Id;
    }

    public override bool Equals(object? obj) => Equals(obj as Timeline);

    public override int GetHashCode() => Id.GetHashCode();

    public static bool operator ==(Timeline? left, Timeline? right) =>
        left is null ? right is null : left.Equals(right);

    public static bool operator !=(Timeline? left, Timeline? right) =>
        !(left == right);

    public override string ToString() =>
        $"Timeline({Name}, {ClipCount} clips, {TotalDuration:F2}s)";
}
